### PR TITLE
Features/pbazzi/rest refactor closing apache connector2

### PR DESCRIPTION
**Pull request by paolobazzi**
Replace ClosingApacheConnector by ScoutApacheConnector implementation

- Adding several HTTP client features like Scout dynamic http proxy selector.
- Bugzilla 572405 

300860, 237883

### CODE REVIEW

**Comment by paolobazzi** [comment]
Jenkins please run tests

**Comment by paolobazzi** [comment]
Jenkins please run tests

File: org.eclipse.scout.rt.rest.jersey.client/src/main/java/org/eclipse/scout/rt/rest/jersey/client/JerseyClientConfigFactory.java [hunk 1]
```diff
@@ -160,13 +95,18 @@ protected void initLoggingFeature(ClientConfig clientConfig) {
...
   /**
-   * Initializes connection provider. Additionally, creates a default connection manager if none has been set previously
-   * in {@code clientConfig}, allowing for consumers to provide their own connection manager.
+   * Initializes connection provider based on configuration.
    */
   protected void initConnectionProvider(ClientConfig clientConfig) {
-    clientConfig.connectorProvider(new ClosingApacheConnectorProvider());
-    if (clientConfig.getProperty(ApacheClientProperties.CONNECTION_MANAGER) == null) {
-      clientConfig.property(ApacheClientProperties.CONNECTION_MANAGER, createConnectionManager(clientConfig));
+    if (Boolean.TRUE.toString().equals(clientConfig.getProperty(RestClientProperties.USE_SCOUT_APACHE_CONNECTOR)) || CONFIG.getPropertyValue(UseScoutApacheConnectorProperty.class)) {
```
Comment (andibur) [inline org.eclipse.scout.rt.rest.jersey.client/src/main/java/org/eclipse/scout/rt/rest/jersey/client/JerseyClientConfigFactory.java]:
```suggestion
    if (Boolean.parseBoolean(clientConfig.getProperty(RestClientProperties.USE_SCOUT_APACHE_CONNECTOR)) || CONFIG.getPropertyValue(UseScoutApacheConnectorProperty.class)) {
```

File: org.eclipse.scout.rt.rest.jersey.test/src/main/java/org/eclipse/scout/rt/rest/jersey/JerseyTestApplication.java [hunk 2]
```diff
@@ -58,17 +59,27 @@ public int getPort() {
     return m_port;
   }
 
+  public int getProxyPort() {
+    return m_proxyPort;
+  }
+
   protected void startInternal() throws Exception {
     m_port = getListenPort();
+    m_proxyPort = getListenPort();
```
Comment (andibur) [inline org.eclipse.scout.rt.rest.jersey.test/src/main/java/org/eclipse/scout/rt/rest/jersey/JerseyTestApplication.java]: It looks like both ports are the same. The method should be renamed to `findUnusedNetworkPort()`.

File: org.eclipse.scout.rt.rest.jersey.test/src/main/java/org/eclipse/scout/rt/rest/jersey/RestClientTestEchoServlet.java [hunk 3]
```diff
@@ -92,10 +102,17 @@ public void sendEchoResponse(HttpServletRequest req, HttpServletResponse resp, i
...
+    String cookieValue = req.getParameter(COOKIE_VALUE);
+    if (cookieValue != null) {
+      resp.addCookie(new Cookie(ECHO_SERVLET_COOKIE, cookieValue));
+    }
+
     if (req.getParameter(EMPTY_BODY) != null) {
       return;
     }
 
+    String body = IOUtility.readString(req.getInputStream(), StandardCharsets.UTF_8.name());
```
Comment (andibur) [inline org.eclipse.scout.rt.rest.jersey.test/src/main/java/org/eclipse/scout/rt/rest/jersey/RestClientTestEchoServlet.java]:
```suggestion
    String body = IOUtility.readStringUTF8(req.getInputStream());
```

File: org.eclipse.scout.rt.rest.jersey.test/src/main/java/org/eclipse/scout/rt/rest/jersey/TestingRestClientConfigFactory.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        return true;
+      }
+
+      /**
+       * Creates a {@link HttpClientConnectionManager} that manages exactly one connection. This limitation helps
+       * finding resource leaks (i.e. leased connections that are never put back to the pool).
+       */
+      @Override
+      protected HttpClientConnectionManager createConnectionManager(Client client, Configuration config, SSLContext sslContext) {
+        m_httpClientConnectionManager = super.createConnectionManager(client, config, sslContext);
```
Comment (andibur) [inline org.eclipse.scout.rt.rest.jersey.test/src/main/java/org/eclipse/scout/rt/rest/jersey/TestingRestClientConfigFactory.java]: It is a bit strange to keep the return value during the initialization of an object that is created by a factory. Maybe you could additionally comment the intention.

**Review by andibur — COMMENTED** [review]
It looks good to me.

- Do you squash the commits because the first one contains some mistakes?
- There are some FIXMEs left.

File: org.eclipse.scout.rt.rest.jersey.client/src/main/java/org/eclipse/scout/rt/rest/jersey/client/JerseyClientConfigFactory.java (hunk 1 above)
Comment (paolobazzi) [reply org.eclipse.scout.rt.rest.jersey.client/src/main/java/org/eclipse/scout/rt/rest/jersey/client/JerseyClientConfigFactory.java]: getProperty() returns Object, but I'll changed it to ``clientConfig.isProperty(...)`` which returns boolean directly

File: org.eclipse.scout.rt.rest.jersey.test/src/main/java/org/eclipse/scout/rt/rest/jersey/JerseyTestApplication.java (hunk 2 above)
Comment (paolobazzi) [reply org.eclipse.scout.rt.rest.jersey.test/src/main/java/org/eclipse/scout/rt/rest/jersey/JerseyTestApplication.java]: Done

File: org.eclipse.scout.rt.rest.jersey.test/src/main/java/org/eclipse/scout/rt/rest/jersey/RestClientTestEchoServlet.java (hunk 3 above)
Comment (paolobazzi) [reply org.eclipse.scout.rt.rest.jersey.test/src/main/java/org/eclipse/scout/rt/rest/jersey/RestClientTestEchoServlet.java]: Done

File: org.eclipse.scout.rt.rest.jersey.test/src/main/java/org/eclipse/scout/rt/rest/jersey/TestingRestClientConfigFactory.java (hunk 4 above)
Comment (paolobazzi) [reply org.eclipse.scout.rt.rest.jersey.test/src/main/java/org/eclipse/scout/rt/rest/jersey/TestingRestClientConfigFactory.java]: Changed implementation back to preserve connection manager as property of the used jersey client instance

**Pull request merged**